src/frontends/lean/elaborator.cpp

```diff
@@ -2884,350 +2869,405 @@ expr elaborator::instantiate_mvars(expr const & e, std::function<bool(expr const
-                            /* field from source */
-                            name base_S_name = *opt_base_S_name;
-                            expr base_src = *mk_base_projections(m_env, src.m_S_name, base_S_name, src.m_e);
-                            expr f = mk_proj_app(m_env, base_S_name, S_fname, base_src);
-                            c_arg = visit(f, none_expr());
-                            expr c_arg_type = infer_type(c_arg);
-                            if (!is_def_eq(c_arg_type, d)) {
-                                format msg =
-                                        format("type mismatch at field '") + format(S_fname) +
-                                        format("' from source '") + pp(src.m_e) + format("'");
-                                msg += pp_indent(mk_pp_ctx(), c_arg);
-                                msg += line() + pp_type_mismatch(c_arg_type, d);
-                                report_or_throw(elaborator_exception(ref, msg));
-                                c_arg = mk_sorry(some_expr(d), ref);
-                            }
-                            field2value.insert(S_fname, c_arg);
-                            return true;
-                        }
-                    }
-                    return false;
-                };
-
-                auto it = std::find(fnames.begin(), fnames.end(), S_fname);
+                c_arg = m_elab.mk_metavar(S_fname.append_before("?"), d, m_ref);
+
+                /* Try to find field value, in the following order:
+                 * 1) implicit field: infer
+                 * 2) explicit value from m_fvalues
+                 * 3) subobject field: recurse
+                 * 4) value from source
+                 * 5) opt/auto param
+                 * 6) `..` catchall: placeholder
+                 */
+                auto it = std::find(m_fnames.begin(), m_fnames.end(), S_fname);
                 if (!is_explicit(binding_info(c_type))) {
-                    /* Implicit field */
-                    if (it != fnames.end()) {
-                        used[it - fnames.begin()] = true;
-                        report_or_throw(elaborator_exception(e, sstream() << "invalid structure value {...}, field '"
-                                                                          << S_fname << "' is implicit and must not be provided"));
+                    /* implicit field */
+                    if (it != m_fnames.end()) {
+                        // avoid redundant error
+                        m_fnames_used.insert(S_fname);
+                        m_elab.report_or_throw(
+                                elaborator_exception(m_e, sstream() << "invalid structure value {...}, field '"
+                                                                  << S_fname
+                                                                  << "' is implicit and must not be provided"));
```

Comment:
Btw, it wasn't clear to me why this restriction actually exists

Reply:
I can't see the context. When I click "View changes", I go to a dead page. Is this still an issue?

Reply:
Here is the line context: https://github.com/leanprover/lean/blob/6da6c143e696dc5a974cbc488dc531abfcca4ed9/src/frontends/lean/elaborator.cpp#L3068
I don't see an implementation reason not to support it

Reply:
I believe this restriction is a leftover, and can be removed. AFAIR, before you wrote the new support for elaboration structure instances, structure instances would be elaborated by first converting the input pre-term into a temporary constructor application, where field values become arguments. This auxiliary pre-term would be constructed without `@` (reason: if `@` is used, then the elaborator uses a very simple elaboration procedure for the application). Thus, we did not have a "place" to put implicit arguments.

Reply:
I see, that makes sense in context. I'll lift it then.